funcx_endpoint/funcx_endpoint/endpoint/interchange.py

```diff
@@ -351,49 +351,79 @@ def _main_loop(self):
                 len(resend_results_messages),
             )
 
-        for results in resend_results_messages:
-            # TO-DO: Republishing backlogged/unacked messages is not supported
-            # until the types are sorted out
-            self.results_outgoing.publish(results)
+            for results in resend_results_messages:
+                # TO-DO: Republishing backlogged/unacked messages is not supported
+                # until the types are sorted out
+                self.results_outgoing.publish(results)
 
         executor = list(self.executors.values())[0]
-        last = time.time()
-
-        while not self._quiesce_event.is_set() and not self._kill_event.is_set():
-            if last + self.heartbeat_threshold < time.time():
-                log.debug("alive")
-                last = time.time()
-
-            self.results_ack_handler.check_ack_counts()
-
-            try:
-                task = self.pending_task_queue.get(block=True, timeout=0.01)
-                log.warning(f"Submitting task : {task}")
-                executor.submit_raw(task)
-            except queue.Empty:
-                pass
-            except Exception:
-                log.exception("Unhandled issue while waiting for pending tasks")
-
-            try:
-                results = self.results_passthrough.get(False, 0.01)
-
-                task_id = results["task_id"]
-                if task_id:
-                    self.results_ack_handler.put(task_id, results["message"])
-                    log.info(f"Passing result to forwarder for task {task_id}")
-
-                # results will be a pickled dict with task_id, container_id,
-                # and results/exception
-                log.warning(f"Publishing message {results['message']}")
-                self.results_outgoing.publish(results["message"])
-                log.warning(f"quiesce_Event : {self._quiesce_event.is_set()}")
-            except queue.Empty:
-                pass
-
-            except Exception:
-                log.exception(
-                    "Something broke while forwarding results from executor "
-                    "to forwarder queues"
-                )
-                continue
+
+        num_tasks_forwarded = 0
+        num_results_forwarded = 0
+
+        def process_pending_tasks():
+            nonlocal num_tasks_forwarded
+            while not (self._quiesce_event.is_set() or self._kill_event.is_set()):
+                try:
+                    task = self.pending_task_queue.get(timeout=1)
```

Comment:
If this is now becoming a separate thread, I think it's important that we set `block=True` so that it will actually sit and wait here. Otherwise, the implementation might actually perform worse than a single thread (GIL contention, etc).

The same applies to the other thread loop.

Reply:
Kevin rightly pointed out to me that `block=True` is the default. I just got confused. 🤷